src/pop_manager.erl

```diff
@@ -0,0 +1,318 @@
+	    {Status, PopManager}
+    end.
+
+
+%% @doc Recieve hashes, request unknown blocks.
+
+on_net_message(SenderAddress, _, send_block_hashes, HashList, PopManager) ->
+    NetSendFn = PopManager#pop_manager.config#pop_manager_config.net_send,
+
+    FilterFn = fun (Hash) -> get_block_status(Hash, PopManager) == unknown end,
+
+    UnknownHashList = lists:filter(FilterFn, HashList),
+
+    NetSendFn(SenderAddress, request_full_blocks, UnknownHashList),
+
+    PopManager;
+
+%% @doc Receive blocks, add them to the structure.
+
+on_net_message(SenderAddress, CurrentTime, send_full_blocks, {Age, BlockList}, PopManager) ->
+    NetSendFn = PopManager#pop_manager.config#pop_manager_config.net_send,
+
+    FoldFn = 
+	fun(Block, PM) ->
+		BlockStatus = get_block_status(Block, PM),
+
+		if (Age == new) and (BlockStatus == unknown) ->
+			NetSendFn(SenderAddress, request_block_hash_range, setup_range_request(Block, PM));
+		   true -> ok
+		end,
+		
+		NewPM = add_block_out_of_order(Block, CurrentTime, PM),
+		NewPM
+	end,
+
+    NewPopManager = lists:foldl(FoldFn, PopManager, BlockList),
+    NewPopManager;
+
+%% @doc Respond to hash range request by computing and sending the range.
+
+on_net_message(SenderAddress, _, request_block_hash_range, {KnownHash1, KnownHash2, UnknownHash}, PopManager) ->
+    NetSendFn = PopManager#pop_manager.config#pop_manager_config.net_send,
+
+    St1 = get_block_status(KnownHash1, PopManager),
+    St2 = get_block_status(KnownHash2, PopManager),
+    
+    if 
+	St1 == in_chain ->
+	    KnownHash = KnownHash1;
+	St2 == in_chain ->
+	    KnownHash = KnownHash2;
+	true ->
+	    KnownHash = maps:get(this_id, pop_chain:get_genisys_block(PopManager#pop_manager.pop_chain))
+    end,
+
+    RequestHashList = compute_block_hash_range(UnknownHash, KnownHash, PopManager),
+
+    NetSendFn(SenderAddress, send_block_hashes, RequestHashList),
+
+    PopManager;
+
+%% @doc Respond to block request by sending all the known requested blocks.
+
+on_net_message(SenderAddress, _, request_full_blocks, HashList, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+    NetSendFn = PopManager#pop_manager.config#pop_manager_config.net_send,
+
+    MapFn = fun (Hash) ->
+		    R1 = pop_chain:find_block_by_id(Hash, PC),
+		    case R1 of
+			{ok, Block} ->
+			    NetSendFn(SenderAddress, send_full_blocks, {old, [Block]});
+			error ->
+			    debug_only:error("unexpected request_full_blocks, unknown hash")
+		    end,
+		    Hash
+	    end,
+
+    lists:map(MapFn, HashList),
+
+    PopManager;
+
+%% @doc Process the recieved transactions.
+
+on_net_message(_, _, send_transactions, TransactionList, PopManager) ->
+    PC = PopManager#pop_manager.pop_chain,
+
+    FoldFn = fun (T, PC0) ->
+		     {_, PC1} = pop_chain:add_transaction(T, PC0),
+		     PC1
+	     end,
+    
+    NewPC = lists:foldl(FoldFn, PC, TransactionList),
+
+    PopManager#pop_manager{pop_chain = NewPC}.
+
+%% @doc Loop that handles messages.
+
+loop(PopManager) ->
```

Comment:
you should turn PopManager into a gen_server

Reply:
loop though is only meant to be used for testing

but yeah I could look into gen_server